### PR TITLE
Fix #104 Retry utils for commonly used retry/repeat patterns

### LINKED ISSUE #104: Add retry utils for commonly used retry patterns of Flux/Mono
Examples of common patterns are DelayUtils in cf-java-client and the retry utils in couchbase-java-client.

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@simonbasle @smaldini Can you take a look? Thank you...

File: reactor-extra/src/main/java/reactor/retry/DefaultRepeat.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+				backoff, jitter, backoffScheduler, onRepeat, applicationContext);
+	}
+
+	@Override
+	public Repeat<T> withBackoffScheduler(Scheduler scheduler) {
+		return new DefaultRepeat<>(repeatPredicate, maxIterations, timeout,
+				backoff, jitter, scheduler, onRepeat, applicationContext);
+	}
+
+	class RepeatFunction implements Function<Flux<Long>, Publisher<Long>> {
```
Comment (simonbasle) [inline reactor-extra/src/main/java/reactor/retry/DefaultRepeat.java]: I wonder if there's a way of still having 1 independent `Publisher` per `DefaultRepeat#apply` but still get rid of that inner `RepeatFunction`. Same for `RetryFunction`.

File: reactor-extra/src/main/java/reactor/retry/DefaultRetry.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+				backoff, jitter, backoffScheduler, onRetry, applicationContext);
+	}
+
+	@Override
+	public Retry<T> withBackoffScheduler(Scheduler scheduler) {
+		return new DefaultRetry<>(retryPredicate, maxIterations, timeout,
+				backoff, jitter, scheduler, onRetry, applicationContext);
+	}
+
+	class RetryFunction implements Function<Flux<Throwable>, Publisher<Long>> {
```
Comment (simonbasle) [inline reactor-extra/src/main/java/reactor/retry/DefaultRetry.java]: I wonder if there's a way of still having 1 independent `Publisher` per `DefaultRetry#apply` but still get rid of that inner `RetryFunction`. Same for `RepeatFunction`.

File: reactor-extra/src/test/java/reactor/retry/RepeatTests.java [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import reactor.test.StepVerifier;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+public class RepeatTests {
```
Comment (simonbasle) [inline reactor-extra/src/test/java/reactor/retry/RepeatTests.java]: if you could come up with a sensible test that ensures storing the `DefaultRepeat` in a variable and reusing it in two `repeatWhen` in parallel doesn't produce weird state due to sharing...

File: reactor-extra/src/main/java/reactor/retry/Backoff.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+package reactor.retry;
+
+import java.time.Duration;
+import java.util.function.Function;
+
+public interface Backoff extends Function<Context<?>, BackoffDelay> {
+
+	static Backoff zero() {
+		return context -> BackoffDelay.ZERO;
```
Comment (rajinisivaram) [inline reactor-extra/src/main/java/reactor/retry/Backoff.java]: Fixed.

File: reactor-extra/src/main/java/reactor/retry/Backoff.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		Duration maxBackoffInterval = maxBackoff != null ? maxBackoff : Duration.ofSeconds(Long.MAX_VALUE);
+		if (maxBackoffInterval.compareTo(firstBackoff) <= 0)
+			throw new IllegalArgumentException("maxBackoff must be >= firstBackoff");
+		return context -> {
+			Duration nextBackoff = firstBackoff.multipliedBy((long) Math.pow(2, (context.iteration() - 1)));
+			return new BackoffDelay(firstBackoff, maxBackoffInterval, nextBackoff);
+		};
+	}
+
+	static Backoff prevTimes3(Duration firstBackoff, Duration maxBackoff) {
```
Comment (rajinisivaram) [inline reactor-extra/src/main/java/reactor/retry/Backoff.java]: Couldn't think of a better method name. So combined the methods to use different parameters instead.

File: reactor-extra/src/main/java/reactor/retry/DefaultRepeat.java (hunk 1 above)
Comment (rajinisivaram) [inline reactor-extra/src/main/java/reactor/retry/DefaultRepeat.java]: We basically just need somewhere to store the previous backoff. That is the only mutable state. Moved that state to DefaultContext and removed the inner class.

File: reactor-extra/src/main/java/reactor/retry/DefaultRetry.java (hunk 2 above)
Comment (rajinisivaram) [inline reactor-extra/src/main/java/reactor/retry/DefaultRetry.java]: Same as above.

File: reactor-extra/src/test/java/reactor/retry/RepeatTests.java (hunk 3 above)
Comment (rajinisivaram) [inline reactor-extra/src/test/java/reactor/retry/RepeatTests.java]: @simonbasle  Thank you for the review. Added a test that uses shared `DefaultRepeat` and `DefaultRetry` from multiple threads.

**Comment by rajinisivaram** [comment]
@simonbasle Thank you for the reviews. I have added javadoc to the new code. Merging now...

**Pull request merged**